pkg/validator/expression_helpers.go

```diff
@@ -46,3 +47,37 @@ func getExpressionSelectors(expr string) ([]string, error) {
 	})
 	return selectors, nil
 }
+
+func getExpressionMetricsNames(expr string) ([]string, error) {
+	promQl, err := parser.ParseExpr(expr)
+	if err != nil {
+		return []string{}, fmt.Errorf("failed to parse expression `%s`: %s", expr, err)
+	}
+	var names []string
+	parser.Inspect(promQl, func(n parser.Node, ns []parser.Node) error {
+		switch v := n.(type) {
+		case *parser.VectorSelector:
+			names = append(names, getMetricNameFromVectorSelector(v))
+		}
+		return nil
+	})
+	return names, nil
+}
+
+func getMetricNameFromVectorSelector(v *parser.VectorSelector) string {
```

Comment:
Actually, as I found out [here](https://github.com/FUSAKLA/promruval/pull/35/files#diff-42b4b15ab9852f1087b8dffb7f60cad21bb156bfff17eff476400cc3f1e6dcf9R432) the metric name used in the vector selector gets always injected also as a LabelMatcher `__name__="foo"`

So it might be enough just to check for the `__name__` selector in the `v.LabelMatchers`?
I suppose creating the new `VectorSelector` is a workaround for this behavior?

Reply:
updated